core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -426,8 +426,16 @@ class ReplicaManager(val config: KafkaConfig,
   def getPartitionAndLeaderReplicaIfLocal(topicPartition: TopicPartition): (Partition, Replica) =  {
     val partitionOpt = getPartition(topicPartition)
     partitionOpt match {
+      case None if metadataCache.contains(topicPartition.topic) =>
+        // The topic exists, but this broker is no longer a replica of it, so we return NOT_LEADER which
+        // forces clients to refresh metadata to find the new location. This can happen, for example,
+        // during a partition reassignment if a produce request from the client is sent to a broker after
+        // the local replica has been deleted.
+        throw new NotLeaderForPartitionException(s"Broker $localBrokerId is not the leader of $topicPartition")
```

Comment:
Should we say that the broker is not one of the partition replicas? Also, maybe we can tweak the existing message below:

```scala
"Leader not local for partition $topicPartition on broker $localBrokerId"
```

It seems like a long winded way to say it. Finally, do we also want to tweak:

```scala
"Partition $topicPartition doesn't exist on $localBrokerId"
```

After this change, it doesn't just mean that the partition doesn't exist in this broker. The topic also doesn't exist according to the Metadata cache. These messages don't get sent to the client, but if we log them, then the additional detail may help.

Thoughts?